### PR TITLE
GeneratorPipeline optimization

### PR DESCRIPTION
**Pull request by Cyclonit**
Because the generator pipeline eats up lots of CPU cycles, I decided to clean it up and get rid of some stuff. QueueProcessor contains lots of expensive operations (regarding its 4 set structure), which are no longer required. Cubes will no longer be skipped during processing, because all required cubes are guaranteed to be available, therefor keeping track of skipped cubes and requeing them is no longer necessary. By using a linked list, adding new work to the processing queues and removing individual entries is improved.

Additionally, I decided to move the queues into GeneratorPipeline and calling CubeProcessor.calculate directly. Thus, the number of convoluted accesses to other objects is reduced.

### CODE REVIEW

**Comment by Barteks2x** [comment]
Eventually I would like to automatically sort the "queue" by distance to closest player every few ticks, would be nice to include that while reworking this part.

File: src/main/java/cubicchunks/worldgen/GeneratorPipeline.java [hunk 1]
```diff
@@ -25,31 +25,33 @@
...
 import java.util.HashMap;
+import java.util.LinkedList;
 import java.util.List;
 import java.util.Map;
+import java.util.Queue;
 
 public class GeneratorPipeline {
 
-	private static final int TickBudget = 40; // ms. There are only 50 ms per tick
+	private static final int MAX_CUBES_PER_TICK = 500;
```
Comment (Barteks2x) [inline src/main/java/cubicchunks/worldgen/GeneratorPipeline.java]: this is way too high. The goal of this is to make the server still work relatively smoothly even while generating chunks.

File: src/main/java/cubicchunks/worldgen/GeneratorPipeline.java [hunk 2]
```diff
@@ -129,129 +126,129 @@ public void generate(Cube cube, GeneratorStage targetStage) {
...
+
+			for (int i = 0; i < this.stages.size(); ++i) {
+				GeneratorStage stage = this.stages.get(i);
+				int processed = processStage(stage);
+				totalProcessed += processed;
+				if (processed > 0) {
+					done = false;
+					break;
+				}
+			}
```
Comment (Barteks2x) [inline src/main/java/cubicchunks/worldgen/GeneratorPipeline.java]: Would be nice to also display "generation rounds" as previously.

**Comment by Barteks2x** [comment]
Aside of my comments, is it ready to merge?

File: src/main/java/cubicchunks/worldgen/GeneratorPipeline.java (hunk 1 above)
Comment (Barteks2x) [inline src/main/java/cubicchunks/worldgen/GeneratorPipeline.java]: optional time limit should also be there

Comment (Cyclonit) [inline src/main/java/cubicchunks/worldgen/GeneratorPipeline.java]: Regarding the maximum, this was the old value and I just stuck with it. I didn't see lots of "can't keep up" messages during testing, so I assumed its fine. What number do you suggest? I guess ultimately some form of automatic adaptation to the server performance might be best. On my rig (which is likely overpowered to serve as a good benchmark) managed ~2500-3000 processed cubes/s.

File: src/main/java/cubicchunks/worldgen/GeneratorPipeline.java (hunk 2 above)
Comment (Cyclonit) [inline src/main/java/cubicchunks/worldgen/GeneratorPipeline.java]: I'll add it back it.

**Comment by Cyclonit** [comment]
How do you intend to do the sorting if there are multiple players? In general, sorting a linked list is fairly cheap, but the performance might take a hit.

**Comment by Barteks2x** [comment]
The idea is that it would be performed only when there are no new cubes added (which implies that players players didn't move enough to load new chunks) and only every few ticks. CubePlayerMap does sorting too, and it's not that bad.

**Comment by Cyclonit** [comment]
I'd say its ready to merge. I'll add documentation with a future pull request.

**Comment by Barteks2x** [comment]
I didn't see that it removed reporting worldgen progress when generating spawnpoint.

**Comment by Barteks2x** [comment]
I removed the merge commit, after it's fixed - submit next pull request.

**Comment by Cyclonit** [comment]
The progress by stages shows the progress since the last report. Therefor, the NAN/s are intended behavior. Do you want the total stats to be logged instead?

**Comment by Barteks2x** [comment]
For me the processing itself didn't seem to work at all, at least for VanillaCubic. And the output didn't really tell me what was going on.

**Comment by Cyclonit** [comment]
I do see VanillaCubic not working and I'm searching for the problem. Apparently some cubes are discarded from the generator queues without being processed...

**Comment by Barteks2x** [comment]
VanillaCubic is just a weird setup to begin with, which means that it's a good test for the dependency system.

This is the only one where the first stage has any dependencies. Cubes at y=1 to 15 depend on cube at y=0, and cube at y=0 depends on cubes at y=1 to 15. This means that generating any cube from y=0 to 15  range requires generating all of them.

Then there is standard lighting stage which required cubes in radius 2. Which means that lighting actually requires cubes cubes between y=0 and  16 for columns in radius 2.

And then there is population stage, in which cubes between y=1 and y=15 depend on cubes bettween (x, 0, z) and (x+1, 0, z+1), and cube at y=0 depends on cubes between y=1 and y=15.

This is all to simulate vanilla chunk dependencies.

**Comment by Barteks2x** [comment]
After commit 276100b95de8c7c41f026041b1531d6040775023 most of these changes are not necessary anymore, but new reporting code may be still useful.

**Pull request merged**